Parse and Deserialize logs from JSON and Print to Console

frextrite opened the pull request:
Parse and deserialize logs when coala is run with `--log-json`

Extension: print these logs to console.

frextrite committed:
CodeInspectionSeverity.java: Add Verbose severity

frextrite committed:
CodeAnalysisConsoleView.java: Implement verbose print method

frextrite committed:
CodeAnalysisConsoleView.java: Add new text attribute for DEBUG

The original DEBUG color used in JB is Blue
coala uses GREEN

frextrite committed:
CodeAnalysisConsoleView.java: Add method to clear console output

frextrite committed:
CodeAnalysisLog.java: Create class to hold log information

frextrite committed:
DeserializeLogFromJson.java: Parse Log JSON

This class is used for parsing log from JSON
and creating log objects

frextrite committed:
CodeAnalysisLogDeserializer.java: Parse JSON and deserialize

The main class responsible for parsing of JSON string
and calling helper methods to deserialize

frextrite committed:
CodeAnalysisLogDeserializerTest.java: Add tests

ksdme commented on src/main/java/io/coala/jetbrains/utils/CodeAnalysisLog.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.Date;
+import java.util.TimeZone;
+
+public class CodeAnalysisLog {
```
This class represents a line in the log, right? If so, please name is appropriately.

frextrite replied:
Each instance represents one log block in the JSON having severity, message, and timestamp of when it was generated. I kept the name similar to `CodeAnalysisIssue` (each instance represents one issue) since both are similar in nature.

ksdme replied:
What does `issue` here refer to? You are not using this to get lint results, are you?

ksdme commented on src/main/java/io/coala/jetbrains/utils/CodeAnalysisLog.java:
```diff
@@ -0,0 +1,69 @@
+    return null;
+  }
+
+  private CodeInspectionSeverity getSeverityLevelFromString(String severity) {
```
This needs to be `static` method. Preferably of also class `CodeAnalysisLogDeserializer` instead.

frextrite replied:
And public also right?

ksdme replied:
That is your call, export it if you think you might need it somewhere otherwise keep external interface clean and simple.

ksdme commented on src/main/java/io/coala/jetbrains/utils/deserializers/CodeAnalysisLogDeserializer.java:
```diff
@@ -0,0 +1,62 @@
+    public List<CodeAnalysisLog> deserialize(JsonElement jsonElement, Type type,
+        JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
+      final JsonObject jsonObject = jsonElement.getAsJsonObject();
+      final JsonArray logsList = jsonObject.get("logs").getAsJsonArray();
```
`logs` literal shoul preferably be a static property.

frextrite replied:
I don't see any benefit of making this `static`. This is just an intermediate list of raw json logs which are processed in the next step.

ksdme replied:
Those kinds of suggestions are generally meant to maintain immunity towards API breakages. If you end assuming the literal in multiple places, you'll have to update them all which isn't desireable.

frextrite replied:
That's right. However, this is the only place I've used `"logs"` since I need to access the property directly from the JSON string. If coala decides to change this in the future, on the string at a single needs to be changed.

frextrite committed:
CodeAnalysisLog.java: Get severity tag

add method to get severity tag according to the severity

make helper methods public static to follow DRY

frextrite committed:
CodeAnalysisLogPrinter.java: Print logs to console

create a new project class for printing logs to the console
in coala's signature style
this is the topmost class in the hierarchy of print abstraction

frextrite committed:
plugin.xml: Add CodeAnalysisLogPrinter as Project Component

frextrite committed:
CodeAnalysisTask.java: Parse logs from JSON

and print to console

frextrite committed:
CodeAnalysisRunner.java: Print log information to console

implement method to automatically show coala
Tool Window when an analysis is started

frextrite commented:
A big PR with 13 commits.

Changes: 
1. Parsing JSON - Created classes `CodeAnalysisLog` (modelling log objects), `DeserializeLogFromJson` (class used to parse JSON and get log objects, helper of gsonbuilder), `CodeAnalysisLogDeserializer` (the main abstraction) and updated corresponding test class `CodeAnalysisLogDeserializerTest`
2. Printing to console - `CodeAnalysisLogPrinter` responsible for printing log objects to the console. Other affected classes: `CodeAnalysisRunner` (print the cmd run statement), `CodeAnalysisTask` is the main calling class for the project, responsible for calling the printer class to print logs.
3. Other changes - `CodeAnalysisConsoleView` Add verbose severity print methods, and create custom colors for `DEBUG` tag (IntelliJ uses blue, coala used Green)

frextrite commented:
Here's a nice screenshot for the same:
![coala_console](https://user-images.githubusercontent.com/29180487/61309441-d4d1b600-a80f-11e9-9e64-9e651d9c2cfe.png)
